File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        HashMap <Character,Integer> map = new HashMap<>();
        int l = 0 ;
        int maxLength = 0;
        for(int r = 0 ; r<n ; r++){
            map.put(s.charAt(r),map.getOrDefault(s.charAt(r),0)+1);
            while(map.get(s.charAt(r))>1){
                map.put(s.charAt(l),map.getOrDefault(s.charAt(l),0)-1);
                l++;
            }
            maxLength = Math.max(maxLength,r-l+1);
        }
        return maxLength ;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxi = 0
        dic={}
        i,j = 0,0
        while j<len(s) :
            ch = s[j]
            dic[ch]=dic.get(ch,0)+1

            while dic[ch]>1:
                dic[s[i]] = dic.get(s[i])-1
                i += 1
            maxi = max(maxi,j-i+1)
            j += 1
             
        return  maxi        